vscode: brand new tmLanguage grammar

Monkatraz opened the pull request:
Replaces the current tmLanguage grammar with a new, feature-rich one. This new grammar makes significant reuse of the existing VSCode HTML grammar. Due to this, and despite boasting a significant amount of new features, it is only ~100 lines longer in its JSON form, as compared to the original.

The source of this grammar is in a new `yaml` file. I have not created a script to automatically compile this file for you, because it's really easy to compile with `js-yaml`. Simply:
```
$: js-yaml syntaxes/svelte.tmLanguage.src.yaml > syntaxes/svelte.tmLanguage.json
```

### The gist:
* The new grammar makes use of scary, undocumented, but actually ancient and stable TML features to significantly reduce the complexity of the grammar. Its `yaml` form is only ~550 lines, and that's with a load of comments.
* It's now stupidly easy to add new embedded languages, as it's just a copy paste of 3 lines.
* Significant reuse of the HTML grammar, meaning that in most cases the scopes of an HTML file will exactly match the scopes of a Svelte file (except for the scopes the Svelte grammar adds, of course).
* I took great effort to get the scope names used to be at least _close to correct_ for most tokens.
* Most of the Svelte-specific markup syntax now has syntax highlighting directly targeting it. For example, the `slot` and component tags have differing attribute highlighting to signify the fact that their attributes are more like object properties.
* The Svelte-specific markup that was addressed by the original grammar has also been improved.
* A few... 'bugs'? were fixed. The biggest 'fix' is probably that now the grammar wraps embedded languages with the `meta.embedded.block.svelte` or `meta.embedded.expression.svelte` scope. This _should_ mean VSCode handles things a little nicer, as it handles that scope explicitly with some of its injected language handling.

There is _a lot_ to talk about with this grammar, so much so I really can't go through it all. I know that's a bit... distasteful for a PR, but I really couldn't make the grammar how I wanted without just remaking the entire thing. I recommend scrolling through the `yaml` source to find out more. I tried my best to make it clear, and I honestly don't think much beyond the `attributes` patterns are that complex.

I do suspect there is problems somewhere with it, but probably nothing significant. If I were to see issues popping up, it's with the Markdown documentation comments and _potentially_ the inline CSS-grammar style attribute. Both can be easily removed or adjusted.

Finally, there is some work left to be done, although this particular PR can be merged. Specifically, the Pug grammar I made should be significantly un-screwy-ified by incorporating it directly within this grammar.  The patterns in this grammar can just be reused directly in Pug now. Additionally, some upcoming or obscure Svelte features may be missing - one that will probably need to be looked at is the RFC13 `--css-var={myCSSVariable}` syntax.

Oh, also, if you do edit the `yaml` source, make sure to turn off your auto-formatter. They're often way too aggressive - some of the unusual code structure I used was specifically to make the grammar way easier to read.

### Gimme pictures:
![image](https://user-images.githubusercontent.com/34875062/98334698-ace80200-1fc0-11eb-89a9-1703878be65d.png)
![image](https://user-images.githubusercontent.com/34875062/98334779-e1f45480-1fc0-11eb-9f6d-27cc16dcf84f.png)
![image](https://user-images.githubusercontent.com/34875062/98334819-fafd0580-1fc0-11eb-87f6-d33ea5aa432e.png)
![image](https://user-images.githubusercontent.com/34875062/98334839-06e8c780-1fc1-11eb-9adc-aef61d6b9f46.png)
![image](https://user-images.githubusercontent.com/34875062/98334873-1cf68800-1fc1-11eb-944f-4465c15ace98.png)
![image](https://user-images.githubusercontent.com/34875062/98335055-78287a80-1fc1-11eb-8fcf-efac288c4a1f.png)
![image](https://user-images.githubusercontent.com/34875062/98335682-a6f32080-1fc2-11eb-87c5-1eca54225b59.png)

#### With VSCode's Dark+ theme:
![image](https://user-images.githubusercontent.com/34875062/98334741-cab56700-1fc0-11eb-9a77-ec4982e904ab.png)

firefish5000 commented:
```jsx
{#each array as item}
```
Itellisense does not fully catch the last character of `item`. Mouseover will highlight `ite` and show `ite` in the tooltip. 
intellisense for usage of `item` works properly, but still shows the name as `ite`

jasonlyu123 commented:
Nice works, But I have something to discuss. 

I think since it's using YAML as a source now maybe we can integrate it into the build step and git-ignore the JSON file, or add documentation about the build step. In both cases, I think it would be nice if there's a convert script in the `package.json`.

And here's some highlighting issuce I noticed:
1. 
```svelte
<a class="{className}" >
```
2. 
```svelte
<a class="btn {className}" {href}>
```
3.
```svelte
<Component on:some-event />
```

Monkatraz commented:
![image](https://user-images.githubusercontent.com/34875062/98353923-41f8f400-1fdd-11eb-952b-1fe1d15e56e6.png)

Okay, I fixed those last two. ...I think. The quotes one I believe I fixed - but Regex can be hard to predict, especially when grammars let inside `begin`-`end` blocks break out of their matched bounds. 

However, that first one is dicey... I assume you mean that the strings are scoped in with the brackets (because normally they're only there for the sake of editors), but in this case it really is a string because it's the `class` attribute.

Honestly, I really don't know how to handle that one. I can't think of an elegant solution - other than a literal list of 'always consider me a string, please' attributes.

Monkatraz commented:
About the `yaml` build step: I avoided making any changes beyond just changing the grammar. However, if a build step were added, we could - erm - *borrow* the [TypeScript tmLanguage](https://github.com/microsoft/TypeScript-TmLanguage) variables script. This would let repetitive elements (e.g. TS var. idents) be a bit less messy. This grammar is pretty light on needing that though.

dummdidumm commented:
Thanks for this awesome rework!

Feedback:
- `<div class="asdasd" {a}></div>` breaks syntax highlighting, every html afterwards is highlighted wrong.
- The attributes on style/script  are highlighted differently, I'd prefer it to be the same as for normal tags
- Same for properties on components or special svelte tags: I prefer them to have the same color as html attribute values, I think that makes it easier to know what to look for, and their semantic meaning ("attribute/prop") is very similar IMHO
- `<script lang="ts" /> <div></div>` the div tags are highlighted wrong when there's at least one attribute a script/style tag and it's self-closed (some do this and reference a src)
- highlighting of subsequent attributes breaks when there's a style attribute: `<div style="width: 20px" class="asd"></div>`

About the build script: I think it would be enough to just have the command you mentioned at the top (js-yaml) as a script inside the package.json, adding js-yaml as a dev dependency and then run that script as part of the build script. Putting the generated file inside `.gitignore` like @jasonlyu123 suggested is a good idea IMO - I just hope that `vsce` does not take the `.gitignore` into account, not publishing the JSON as a result.

Monkatraz commented:
Okay, addressed all of those bugs. I also removed the quote interpolation wrapping thingy, it was just kind of messy. Let me do that build script thingy now. Make sure to double check this, because I have no idea how the deploy script handles everything.

Monkatraz commented:
Should be good to go now if there isn't anything glaringly obvious I missed

dummdidumm reviewed (approved):
Thanks for the quick fixes! Tested this as well as packaging and everything works. This is good from my side now.

dummdidumm commented on packages/svelte-vscode/package.json:
```diff
@@ -4,7 +4,9 @@
     "main": "dist/src/extension.js",
     "scripts": {
-        "build": "tsc -p ./",
+        "build:grammar": "npx js-yaml syntaxes/svelte.tmLanguage.src.yaml > syntaxes/svelte.tmLanguage.json",
```
npx is not needed here
```suggestion
        "build:grammar": "js-yaml syntaxes/svelte.tmLanguage.src.yaml > syntaxes/svelte.tmLanguage.json",
```

dummdidumm reviewed (changes requested):
Found three more bugs:
- `<div></div><Component prop={true}></Component>` <- if there's no space between the closing and opening tag, the Component tag loses all highighlting. Also happens for void tags, for example `<input /><Component></Component>`
- `<div style="{someVariable}"></div>` <- someVariable has wrong token, it's treated as a css token but should be a variable
- `<div style="background: {color} display: flex;"></div>` <- tokenization breaks because of moustache tag

---> To solve the last two I think it might be better to turn off CSS highlighting within style attributes for now

Monkatraz commented:
Okay, got that all done. That last one was confusing me, but I decided to take a sleep, do something I thought I had already tried before, and doing that fixed it on my end. I'm just... not gonna think about it!

dummdidumm commented:
The "opening tag right after closing tag"-bug is still there. For these it does not work:
- `<input /><Component></Component>`
- `<input /><div></div>`
- `<slot /><Component></Component>`
- `<Component></Component><Component></Component>`
- `<div></div><p></p>`

I guess this is some general problem with a regex, looking at the token it seems that the HTML grammar is not injected in this case, `punctuation.definition.tag.begin.html` is always missing when looking at the token of the `<` after the `/>`

dummdidumm commented on packages/svelte-vscode/syntaxes/svelte.tmLanguage.src.yaml:
```diff
@@ -0,0 +1,552 @@
+  tags-void:
+    begin: (?=<(area|base|br|col|embed|hr|img|input|link|meta|param|source|track|wbr)\b)
+    name: meta.tag.void.svelte
+    end: (?!<)(?<=/?>)
```
I think this has to be a negative lookbehind
```suggestion
    end: (?<!<)(?<=/?>)
```

dummdidumm commented on packages/svelte-vscode/package.json:
```diff
@@ -4,7 +4,9 @@
+        "build:ts": "tsc -p ./",
+        "build": "npm run build:ts && npm run build:grammar",
         "vscode:prepublish": "npm run build && npm prune --production",
         "watch": "tsc -w -p ./",
```
Might make sense to add it to watch mode, too.

```suggestion
        "watch": "npm run build:grammar && tsc -w -p ./",
```

Monkatraz commented:
What's happening is that the 'wrap-around' pattern is entirely correct - but when the pattern for getting the start node of the tag block is checked, it will, due to archaic, stupid, absolutely utterly pointless reasons, be allowed to break out of the string of characters it is nested in because it's looking for the last possible `>` character it can find. I have a feeling if I plug this hole, I add a new one somewhere else.

I'll try a few more normal fixes first, but if I don't get it entirely fixed, I have another strategy I can try where I use `match` to get contiguous nodes.

Monkatraz commented:
#### Status Report:

Okay, so you folks discovered a major blocking issue. The issue is that in order for the grammar to practically reuse the HTML grammar, it has to 'wrap-around' nodes. This 'wrap-around' cannot actually use a proper regex wraparound - it instead consumes the character after the end of a tag. This effect is incredibly non-intuitive, and you literally have to work out in your head how the scanner is parsing each character at a time.

The reason it consumes that extra end character is because the injected HTML grammar will consume the ">" in e.g. `<div>`. Thus, my grammar has to end one character afterwards. I cannot have multiple scopes consume the same string - and if there is a way, it's incredibly hacky and stupid. I have tried a lot of techniques to get this to work - and have gotten close - but not close enough.

This is purely due to _multiple_ Textmate limitations - pushing and popping scopes is done implicitly by the scanner, injected languages are allowed to break out of their injected scope, etc. There is a lot of really frustrating technical details here, and I'm not even fully sure how the hell `vscode-textmate` is parsing this.

So what's the solution? Well, the obvious one. I'm just gonna not reuse the HTML grammar. I'll remake the bits of it I need. This is somewhat unfortunate, but I've realized that almost no extension actually makes use of HTML's tag metadata scopes, because depending on Textmate is a terrible, terrible idea. Everything uses the LSP, thankfully.

dummdidumm commented:
Sounds reasonable. Highlighting wise there don't need to be that many different tokens I guess - one for the tag, one for attributes, one for attribute values.

Monkatraz commented:
![image](https://user-images.githubusercontent.com/34875062/98874522-a67dde00-2437-11eb-82b1-b096f072e752.png)

Alright - this is what I've got working. It no longer depends on the HTML grammar in any capacity, which is kind of sad, but oh well. The way I've implemented this means that the grammar should be fairly easy to extend if needed.

Monkatraz commented:
Oops - found an issue with custom langs. crashing the client - so don't merge yet.

dummdidumm reviewed (approved):
Looks good to me now, thanks a lot!
@jasonlyu123 could you also take one more look?

jasonlyu123 commented:
found another one
```
{#each showGroups as [key, items] (key)}

{/each}
```

dummdidumm commented:
I think the problem is that there is no color specified for the tokens. Some themes seem to handle it better than others. How to work around though... No idea. At least for tags and attributes we could use html tokens maybe.

Monkatraz commented:
> I think the problem is that there is no color specified for the tokens. Some themes seem to handle it better than others. How to work around though... No idea. At least for tags and attributes we could use html tokens maybe.

I made a conscious choice when I changed everything to `.svelte`. When you're making a theme, you're really not supposed to use `.html` or any other language suffix. You're meant to only use them when you want to add language specific syntax highlighting features. Just to make sure, I took comparison screenshots when I made everything `.svelte` and on VSCode's native Dark+ theme, literally nothing changes.

My issue with making anything `.html` is what is supposed to be `.html` and what is supposed to be `.svelte`? I suppose we could use `.htmlx` for accuracy - but that introduces the exact same problem.

dummdidumm commented:
I went through all preinstalled color themes and compared them side-by-side with the old. The new one looks better in all of them 👍 Which highlighting did not work for you that worked before @jasonlyu123 ?
If found one more small bug though: 
```
{#if asd}
  asdddddddddddddddd
  dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd
{:else if asd}dddddd{/if}
```
The `if` of `else if` is not highlighted.

jasonlyu123 commented:
Dark (Visual Studio) when 
```
<Foo on:click let:something></Foo>
```
`click` and `something` is not highlighted anymore. I can still differentiate it, just don't know if that's ok for other people.

dummdidumm commented:
That's fine with me, it's the same for components (I remember someone opening an issue about that, [and you recommended to add a color to some setting](https://github.com/sveltejs/language-tools/tree/master/docs#component-syntax-highlighting-does-not-work) -- maybe we should enhance that troubleshooting docs section a little with the other tokens?). Also while not highlighted, it's still in a different color than the other things so it "jumps to the eye" which okay for me.

jasonlyu123 commented:
Yeah, I'm concerning about this because of that issue. Let's just enhance with other tokens like you suggested.

Monkatraz commented:
![image](https://user-images.githubusercontent.com/34875062/99062177-cdcfca80-255f-11eb-9f9b-3c542b2bc7cf.png)
In this example, I've chosen the scope `entity.name.type` because in JavaScript or TypeScript, this scope fits best with JavaScript built-in _classes_, like `Object` and `Promise`. This scope also fits in well with events and well, "types" of things, like what event type you want to listen to. With themes that actually style this scope, I think this choice will be consistent with JS code.

![image](https://user-images.githubusercontent.com/34875062/99062277-fd7ed280-255f-11eb-8aeb-7b8605dc9efb.png)
In this example, `variable.parameter` is used because um... well sometimes you just have to pick something. I _interpret_ `let:` as a sort of parameter definition for a function, so that's why I chose that scope. This is all very easily adjusted in the grammar - but really it's just that a lot of themes do not color variables.

Also, the syntax highlighting in Svelte would really benefit from semantic tokens in embedded expressions  (and in script tags) - my personal theme has rich coloring for semantic tokens and I'd love to see it in Svelte.

Also - I'll get onto that bug now. I pray for the end to be near, as I think I'm starting to see `yaml` in my dreams.

dummdidumm commented:
Is it possible to assign multiple scopes to one token? So that the editor could pick the first one he has a color for? If so, we maybe could investigate (I can do this, to save you some from nightmares about yaml) to add some of these for more consistent highlighting.
For example I just noticed that `transition:...` is colored differently than`on:...` and I would prefer them to be the same color. (We now enter opinion land)

Monkatraz commented:
Yes - you can do that. Although, it's just stacking the scopes on the token. If a theme has some unusual scope selectors (for example, `source.css variable` is common), it might color strangely. My _personal opinion_ is that the scopes should just be what seems most correct and let user-land themes color them however bizarrely they want.

Because I knew this would either be 1. contentious or 2. have to be adjusted, I made `keyword` like bits of syntax easily adjustable. Here is the patterns for adjusting attribute directives:
![image](https://user-images.githubusercontent.com/34875062/99064918-db874f00-2563-11eb-99f3-5ea32767d037.png)

If you want to assign multiple scopes, you simply add another scope to the `name` field. A (strange) example would be this:
```
name: variable.function.svelte entity.name.type.svelte
```
The last scope has the highest priority, or is "on the top of the stack", or however you'd like to feel about it.

`<opinion-land>`
I made the "animation" directives have their own unique keyword scope because I simply felt like grouping them together. I felt that kind of... made more sense? You could tell from glancing at an element what directives were about more logical things and what was more visual. It is a trade-off in visual "busyness" though.
`</opinion-land>`

dummdidumm commented:
Okay, so we could stack scopes but if only the top most gets the color then nothing is won. So we'll just leave it as is now. With a more fine-graned scoping people are able to fine-tine their colors better, which is good.

Alright, I'll merge now. Massive thanks for all your work and the time to fix all the edge cases 👍

Monkatraz commented:
Well, to be correct (and honest), the top most scope has _priority_ but if nothing in the theme selects it, a lower scope will be used to color instead.

Also, no problem. It was actually pretty fun making this grammar, although I didn't expect it to have so many edge cases!

<s>also I have way too much free-time due to current events lol</s>

dummdidumm commented:
Found one more bug (writing it here because it fits best, even if already merged):

- highlighting breaks for empty script/style/template tags, for example `<script></script>`. In reality this will not occur often though I guess because most people will have something inbetween.

dummdidumm commented:
Another one:

`<a xlink:href="https://www.adorno-gymnasium.de">` <- `in` is highlighted, probably because the `in/out/transition` scope is matched when it shouldn't.

Monkatraz commented:
*rubs temples*

Feel free to keep stockpiling the bugs in here - I'll make a PR fixing them in a little bit, probably.
